node/network/src/nightshade_protocol.rs

```diff
@@ -0,0 +1,97 @@
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::time::Duration;
+
+use futures::future;
+use futures::future::Future;
+use futures::sink::Sink;
+use futures::stream::Stream;
+use futures::sync::mpsc::{channel, Receiver, Sender};
+use log::{warn, info, error};
+
+use chain::ChainPayload;
+use nightshade::nightshade::AuthorityId;
+use nightshade::nightshade_task::Gossip;
+use primitives::hash::CryptoHash;
+use primitives::hash::hash_struct;
+use primitives::network::PeerInfo;
+use primitives::serialize::{Decode, Encode};
+use primitives::types::AccountId;
+
+use crate::peer::PeerMessage;
+use crate::peer_manager::PeerManager;
+
+/// Spawn network task that process incoming and outgoing gossips for nightshade consensus
+///
+/// Args:
+pub fn spawn_consensus_network(
+    pm: PeerManager,
+    inc_msg_rx: Receiver<(CryptoHash, Vec<u8>)>,
+    inc_gossip_tx: Sender<Gossip<ChainPayload>>,
+    out_gossip_rx: Receiver<Gossip<ChainPayload>>,
+    authority_map: HashMap<AuthorityId, AccountId>,
+) {
+    // Spawn a task that decodes incoming messages and places them in the corresponding channels.
+    let task = inc_msg_rx.for_each(move |(_, data)| {
+        match Decode::decode(&data) {
+            Ok(gossip) => forward_msg(inc_gossip_tx.clone(), gossip),
+            Err(e) => warn!(target: "network", "Error decoding gossip: {}", e)
+        };
+        future::ok(())
+    });
+    tokio::spawn(task);
+
+    // Spawn a task that encodes and route outgoing gossips to receiver.
+    let task = out_gossip_rx.for_each(move |g| {
+        info!("Sending gossip: {} -> {}", g.sender_id, g.receiver_id);
+        let receiver_channel = authority_map.get(&g.receiver_id).and_then(|acc_id| {
+            pm.get_account_channel(acc_id.clone())
+        });
+
+        if let Some(ch) = receiver_channel {
+            let data = Encode::encode(&g).unwrap();
+            forward_msg(ch, PeerMessage::Message(data));
+        } else {
+            error!("Channel not found!");
+        }
+        future::ok(())
+    });
+    tokio::spawn(task);
+}
+
+fn forward_msg<T>(ch: Sender<T>, el: T)
+    where T: Send + 'static,
+{
+    let task =
+        ch.send(el).map(|_| ()).map_err(|e| warn!(target: "network", "Error forwarding {}", e));
+    tokio::spawn(task);
+}
+
+// TODO: Document this function
+pub fn start_peer(
```

Comment:
This function breaks the encapsulation of the low-level network details. It creates PeerManager, passes it to the code that starts alphanet (breaking the encapsulation of the PeerManager), alphanet waits until there are enough peers, then passes PeerManager back to the network layer and starts the consensus.

Instead it should follow the same setup as existing testnet and devnet where initialization of coroutines happens without waiting for special events, like sufficient number of connected peers.